File: dart-driver/lib/ecf_client.dart
// Cliente HTTP para el API ecf-service.
//
// Métodos públicos:
//   - login(usuario, clave)
//   - enviaEcf(rnc, portal, json)
//   - enviaEcfDesde(documento, portal)
//   - consultaEstado(comprobantes)
//   - descargaXmls(fechaDesde, fechaHasta, {tipos})
//
// Internamente, cada método llama _dispatch(fnName, args) que:
//   1. Llama procesar(estadoJson) del motor estático.
//   2. Si http → shell ejecuta POST y alimenta respuesta al motor.
//   3. Si done → devuelve result (y guarda newToken si vino).
//   4. Si fail → lanza EcfApiError.

import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import 'documento_ecf.dart';
import 'motor.dart';

export 'documento_ecf.dart';

// ---------------------------------------------------------------------------
// Tipos públicos
// ---------------------------------------------------------------------------

/// Excepción cuando el API o el motor devuelve un fallo.
class EcfApiError implements Exception {
  final String code;
  final Map<String, dynamic> data;
  final int? statusCode;
  EcfApiError(this.code, [this.data = const {}, this.statusCode]);

  @override
  String toString() =>
      'EcfApiError(code=$code'
      '${statusCode != null ? ', status=$statusCode' : ''}'
      '${data.isNotEmpty ? ', data=$data' : ''})';
}

/// Archivo recibido en una respuesta (p.ej. ZIP de XMLs en `descargaXmls`).
class EcfFile {
  final String fileName;
  final Uint8List bytes;
  EcfFile({required this.fileName, required this.bytes});
}

/// Resultado de [EcfClient.descargaXmls]: data + archivos.
class EcfDescargaXmlsResult {
  final Map<String, dynamic> data;
  final List<EcfFile> files;
  EcfDescargaXmlsResult({required this.data, required this.files});
}

class _DispatchResult {
  final Map<String, dynamic> result;
  final List<EcfFile> files;
  _DispatchResult({required this.result, required this.files});
}

// ---------------------------------------------------------------------------
// EcfClient
// ---------------------------------------------------------------------------

/// Cliente HTTP para `ecf-service`. Toda la lógica concreta vive en [motor.dart].
class EcfClient {
  final String baseUrl;
  String? _token;
  final http.Client _http;
  final Duration timeout;

  EcfClient({
    String baseUrl = 'https://ecf-service.vicortiz.com',
    String? token,
    Duration timeout = const Duration(seconds: 60),
    http.Client? httpClient,
  }) : baseUrl = baseUrl.replaceAll(RegExp(r'/+$'), ''),
       _token = token,
       _http = httpClient ?? http.Client(),
       timeout = timeout;

  /// Token actual (incluye prefijo `Bearer `). Disponible tras [login].
  String? get token => _token;

  /// Limpia el token (para forzar re-login).
  void clearToken() => _token = null;

  /// Cierra el cliente HTTP subyacente.
  void close() => _http.close();

  // ===========================================================================
  // API pública
  // ===========================================================================

  /// Autentica usuario y guarda token interno. Devuelve `data` del login.
  Future<Map<String, dynamic>> login(
    String usuario,
    String clave, {
    String app = 'ecf',
  }) async {
    final r = await _dispatch('login', {
      'usuario': usuario,
      'clave': clave,
      'app': app,
    });
    return r.result;
  }

  /// Envía un e-CF a la DGII pasando el payload DGII completo (`json`).
  Future<Map<String, dynamic>> enviaEcf({
    required String rnc,
    required String portal,
    required Map<String, dynamic> json,
  }) async {
    final r = await _dispatch('enviaEcf', {
      'rnc': rnc,
      'portal': portal,
      'json': json,
    });
    return r.result;
  }

  /// Envía un [DocumentoEcf] (formato espejo de cursores Fox).
  Future<DocumentoEcf> enviaEcfDesde(
    DocumentoEcf documento, {
    required String portal,
  }) async {
    final encfVal = documento.encf?.trim() ?? '';
    final fiscalVal = documento.fiscal.trim();
    if (encfVal.isNotEmpty && fiscalVal.isNotEmpty) {
      final prefijoCorrecto = 'E$fiscalVal';
      if (!encfVal.startsWith(prefijoCorrecto)) {
        throw ArgumentError(
          'eNCF "$encfVal" no coincide con TipoeCF $fiscalVal '
          '(prefijo esperado: $prefijoCorrecto). DGII error 75.',
        );
      }
    }
    final r = await _dispatch('enviaEcfDesdeDoc', {
      'documento': documento.toMap(),
      'portal': portal,
    });
    final res = r.result;
    documento.encf = (res['encf'] as String?) ?? documento.encf;
    documento.estado = res['estado'] as String?;
    documento.estadoDescripcion = res['estado_descripcion'] as String?;
    documento.codigoSeguridad = res['codigo_seguridad'] as String?;
    documento.fechaFirma = res['fecha_firma'] as String?;
    documento.timbre = res['timbre'] as String?;
    final secU = res['secuencia_utilizada'];
    if (secU is bool) {
      documento.secuenciaUtilizada = secU;
    } else if (secU is num) {
      documento.secuenciaUtilizada = secU.toInt() != 0;
    }
    documento.momento = res['momento'] as String?;
    documento.respuestaMensajes = res['respuesta_mensajes'] as String?;
    return documento;
  }

  /// Consulta estado de e-NCFs (máx 100).
  Future<List<dynamic>> consultaEstado(List<String> comprobantes) async {
    final r = await _dispatch('consultaEstado', {
      'comprobantes': comprobantes,
    });
    return (r.result['result'] as List?) ?? const [];
  }

  /// Descarga ZIP con XMLs de un rango (formato YYYY-MM-DD).
  Future<EcfDescargaXmlsResult> descargaXmls({
    required String fechaDesde,
    required String fechaHasta,
    List<String>? tipos,
  }) async {
    final r = await _dispatch('descargaXmls', {
      'fecha_desde': fechaDesde,
      'fecha_hasta': fechaHasta,
      if (tipos != null && tipos.isNotEmpty) 'tipos': tipos,
    });
    return EcfDescargaXmlsResult(data: r.result, files: r.files);
  }

  // ===========================================================================
  // Trampolín shell ↔ motor
  // ===========================================================================

  Future<_DispatchResult> _dispatch(
    String fnName,
    Map<String, dynamic> args,
  ) async {
    Map<String, Object?>? lastResp;
    List<EcfFile> lastFiles = const [];
    int step = 0;

    while (true) {
      final estado = <String, Object?>{
        'fnName': fnName,
        'args': args,
        'token': _token,
        'step': step,
        'lastResp': lastResp,
      };
      final stepJson = procesar(jsonEncode(estado));
      final stepMap = jsonDecode(stepJson) as Map<String, dynamic>;
      final kind = stepMap['kind'];

      if (kind == 'done') {
        final result =
            (stepMap['result'] as Map?)?.cast<String, dynamic>() ??
                <String, dynamic>{};
        final newToken = stepMap['newToken'] as String?;
        if (newToken != null && newToken.isNotEmpty) _token = newToken;
        return _DispatchResult(result: result, files: lastFiles);
      }

      if (kind == 'fail') {
        final code =
            (stepMap['code'] as String?) ?? 'motor.error_desconocido';
        final dataErr =
            (stepMap['data'] as Map?)?.cast<String, dynamic>() ?? {};
        throw EcfApiError(code, dataErr);
      }

      if (kind == 'http') {
        final endpoint = stepMap['endpoint'] as String;
        final data =
            (stepMap['data'] as Map?)?.cast<String, dynamic>() ??
                <String, dynamic>{};
        final useToken = stepMap['useToken'] != false;
        step = (stepMap['step'] as num?)?.toInt() ?? (step + 1);

        final respFull = await _httpPost(endpoint, data, useToken: useToken);
        lastFiles = _parseFiles(respFull);
        lastResp = {
          'ok': respFull['ok'],
          'message': respFull['message'],
          'data': respFull['data'],
        };
        continue;
      }

      throw EcfApiError('motor.kind_desconocido', {'kind': kind?.toString()});
    }
  }

  // ===========================================================================
  // HTTP de bajo nivel
  // ===========================================================================

  Future<Map<String, dynamic>> _httpPost(
    String endpoint,
    Map<String, dynamic> data, {
    required bool useToken,
  }) async {
    final url = Uri.parse('$baseUrl/');
    final body = jsonEncode({'request': endpoint, 'data': data});
    final headers = <String, String>{'Content-Type': 'application/json'};
    if (useToken && _token != null) headers['Authorization'] = _token!;

    late http.Response resp;
    try {
      resp = await _http
          .post(url, headers: headers, body: body)
          .timeout(timeout);
    } catch (e) {
      throw EcfApiError('http.error', {'detail': e.toString()});
    }

    Map<String, dynamic> out;
    try {
      out = jsonDecode(resp.body) as Map<String, dynamic>;
    } catch (_) {
      throw EcfApiError(
        'respuesta_no_json',
        {
          'text': resp.body.length > 500
              ? resp.body.substring(0, 500)
              : resp.body,
        },
        resp.statusCode,
      );
    }

    if (out['ok'] != true) {
      final msg = (out['message'] as String?) ?? 'error_desconocido';
      final dataErr = (out['data'] as Map?)?.cast<String, dynamic>() ?? {};
      throw EcfApiError(msg, dataErr, resp.statusCode);
    }
    return out;
  }

  List<EcfFile> _parseFiles(Map<String, dynamic> respFull) {
    final raw = respFull['files'];
    if (raw is! List) return const [];
    final out = <EcfFile>[];
    for (final item in raw) {
      if (item is! Map) continue;
      final m = Map<String, Object?>.from(item);
      final name = (m['fileName'] as String?)?.trim() ?? '';
      final b64Raw = (m['content'] as String?) ?? '';
      if (name.isEmpty || b64Raw.isEmpty) continue;
      final b64 = b64Raw.replaceAll(RegExp(r'\s'), '');
      out.add(EcfFile(fileName: name, bytes: base64Decode(b64)));
    }
    return out;
  }
}

File: python-driver/src/chalona_driver/ecf_client.py
"""Cliente HTTP para ecf-service. Toda la lógica de comunicación vive en motor.py.

Métodos públicos:
  - login(usuario, clave)
  - envia_ecf(rnc, portal, json_doc)
  - envia_ecf_desde_doc(documento, portal)
  - consulta_estado(comprobantes)
  - descarga_xmls(fecha_desde, fecha_hasta, tipos?)

Internamente, cada método llama _dispatch(fn_name, args) que:
  1. Llama procesar(estado_json) del motor estático.
  2. Si http → shell ejecuta POST y alimenta respuesta al motor.
  3. Si done → devuelve result (y guarda newToken si vino).
  4. Si fail → lanza EcfApiError.
"""
import json
import urllib.error
import urllib.request
from typing import Any, Optional

from .motor import procesar


class EcfApiError(Exception):
    def __init__(self, code: str, data: Optional[dict] = None, status_code: Optional[int] = None):
        self.code = code
        self.data = data or {}
        self.status_code = status_code

    def __str__(self) -> str:
        parts = [f'EcfApiError(code={self.code}']
        if self.status_code is not None:
            parts.append(f', status={self.status_code}')
        if self.data:
            parts.append(f', data={self.data}')
        return ''.join(parts) + ')'


class EcfClient:
    def __init__(
        self,
        base_url: str = 'https://ecf-service.vicortiz.com',
        token: Optional[str] = None,
        timeout: int = 60,
    ):
        self._base_url = base_url.rstrip('/')
        self._token = token
        self._timeout = timeout

    @property
    def token(self) -> Optional[str]:
        return self._token

    def clear_token(self) -> None:
        self._token = None

    def login(self, usuario: str, clave: str, app: str = 'ecf') -> dict:
        return self._dispatch('login', {'usuario': usuario, 'clave': clave, 'app': app})

    def envia_ecf(self, rnc: str, portal: str, json_doc: dict) -> dict:
        return self._dispatch('enviaEcf', {'rnc': rnc, 'portal': portal, 'json': json_doc})

    def envia_ecf_desde_doc(self, documento: dict, portal: str) -> dict:
        return self._dispatch('enviaEcfDesdeDoc', {'documento': documento, 'portal': portal})

    def consulta_estado(self, comprobantes: list) -> list:
        r = self._dispatch('consultaEstado', {'comprobantes': comprobantes})
        return r.get('result') or []

    def descarga_xmls(
        self,
        fecha_desde: str,
        fecha_hasta: str,
        tipos: Optional[list] = None,
    ) -> dict:
        args: dict = {'fecha_desde': fecha_desde, 'fecha_hasta': fecha_hasta}
        if tipos:
            args['tipos'] = tipos
        return self._dispatch('descargaXmls', args)

    # ---------------------------------------------------------------------------
    # Trampolín shell ↔ motor
    # ---------------------------------------------------------------------------
    def _dispatch(self, fn_name: str, args: dict) -> dict:
        last_resp: Optional[dict] = None
        step = 0

        while True:
            estado = {
                'fnName': fn_name,
                'args': args,
                'token': self._token,
                'step': step,
                'lastResp': last_resp,
            }
            step_json = procesar(json.dumps(estado))
            step_map: dict = json.loads(step_json)
            kind = step_map.get('kind')

            if kind == 'done':
                result = step_map.get('result') or {}
                new_token = step_map.get('newToken')
                if new_token:
                    self._token = new_token
                return result

            if kind == 'fail':
                code = step_map.get('code') or 'motor.error_desconocido'
                data_err = step_map.get('data') or {}
                raise EcfApiError(code, data_err)

            if kind == 'http':
                endpoint: str = step_map['endpoint']
                data: dict = step_map.get('data') or {}
                use_token: bool = step_map.get('useToken', True)
                step = step_map.get('step', step + 1)

                resp_full = self._http_post(endpoint, data, use_token=use_token)
                last_resp = {
                    'ok': resp_full.get('ok'),
                    'message': resp_full.get('message'),
                    'data': resp_full.get('data'),
                }
                continue

            raise EcfApiError('motor.kind_desconocido', {'kind': str(kind)})

    # ---------------------------------------------------------------------------
    # HTTP de bajo nivel
    # ---------------------------------------------------------------------------
    def _http_post(self, endpoint: str, data: dict, *, use_token: bool) -> dict:
        url = f'{self._base_url}/'
        body = json.dumps({'request': endpoint, 'data': data}).encode('utf-8')
        headers: dict = {'Content-Type': 'application/json'}
        if use_token and self._token:
            headers['Authorization'] = self._token

        req = urllib.request.Request(url, data=body, headers=headers, method='POST')
        raw = ''
        status_code = 200

        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                raw = resp.read().decode('utf-8')
                status_code = resp.status
        except urllib.error.HTTPError as e:
            try:
                raw = e.read().decode('utf-8')
            except Exception:
                raise EcfApiError('http.error', {'detail': f'HTTP {e.code}'}, e.code)
            status_code = e.code
        except urllib.error.URLError as e:
            raise EcfApiError('http.error', {'detail': str(e.reason)})
        except Exception as e:
            raise EcfApiError('http.error', {'detail': str(e)})

        try:
            out: Any = json.loads(raw)
        except Exception:
            raise EcfApiError('respuesta_no_json', {'text': raw[:500]}, status_code)

        if not isinstance(out, dict) or out.get('ok') is not True:
            msg = (out.get('message') if isinstance(out, dict) else None) or 'error_desconocido'
            data_err = (out.get('data') if isinstance(out, dict) else None) or {}
            raise EcfApiError(msg, data_err, status_code)

        return out
